File: Drone.Autonomy/Triggers/EventBus.cs
namespace Drone.Autonomy;

public class EventBus
{
    private readonly List<(string? EventType, Func<DroneEvent, Task> Handler)> _handlers = new();
    private readonly object _lock = new();

    /// <summary>Subscribe to all events.</summary>
    public void Subscribe(Func<DroneEvent, Task> handler)
    {
        lock (_lock) { _handlers.Add((null, handler)); }
    }

    /// <summary>Subscribe only to events of a specific type.</summary>
    public void Subscribe(string eventType, Func<DroneEvent, Task> handler)
    {
        lock (_lock) { _handlers.Add((eventType, handler)); }
    }

    /// <summary>Unsubscribe a handler.</summary>
    public void Unsubscribe(Func<DroneEvent, Task> handler)
    {
        lock (_lock) { _handlers.RemoveAll(h => h.Handler == handler); }
    }

    /// <summary>Publish an event to all matching handlers.</summary>
    public async Task PublishAsync(DroneEvent evt)
    {
        (string? EventType, Func<DroneEvent, Task> Handler)[] snapshot;
        lock (_lock) { snapshot = _handlers.ToArray(); }
        foreach (var (eventType, handler) in snapshot)
        {
            if (eventType != null && eventType != evt.Type) continue;
            try { await handler(evt); }
            catch { /* handler errors are non-fatal */ }
        }
    }
}

public class DroneEvent
{
    public string Type { get; }
    public object Data { get; }
    public DateTime Timestamp { get; } = DateTime.UtcNow;

    public DroneEvent(string type, object data) { Type = type; Data = data; }
}

public static class DroneEventTypes
{
    public const string MessageReceived = "MessageReceived";
    public const string FileChanged = "FileChanged";
    public const string ScreenChanged = "ScreenChanged";
    public const string ProcessStarted = "ProcessStarted";
    public const string ProcessStopped = "ProcessStopped";
    public const string SystemAlert = "SystemAlert";
    public const string SystemMetrics = "SystemMetrics";
    public const string ScheduledTask = "ScheduledTask";
}

File: rust/crates/drone-autonomy/src/event_bus.rs
//! Event bus — pub/sub for drone events.

use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use std::sync::Arc;
use tokio::sync::RwLock;

/// Drone event types.
pub struct DroneEventTypes;

impl DroneEventTypes {
    pub const MESSAGE_RECEIVED: &'static str = "message_received";
    pub const FILE_CHANGED: &'static str = "file_changed";
    pub const SYSTEM_METRICS: &'static str = "system_metrics";
    pub const SYSTEM_ALERT: &'static str = "system_alert";
    pub const PROCESS_STARTED: &'static str = "process_started";
    pub const PROCESS_STOPPED: &'static str = "process_stopped";
    pub const SCHEDULED_TASK: &'static str = "scheduled_task";
}

/// A drone event.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DroneEvent {
    pub event_type: String,
    pub data: JsonValue,
}

/// Event handler function type.
type EventHandler = Box<dyn Fn(DroneEvent) -> std::pin::Pin<Box<dyn std::future::Future<Output = ()> + Send>> + Send + Sync>;

/// Async event bus for pub/sub.
pub struct EventBus {
    handlers: Arc<RwLock<Vec<EventHandler>>>,
}

impl EventBus {
    pub fn new() -> Self {
        Self { handlers: Arc::new(RwLock::new(Vec::new())) }
    }

    /// Subscribe to all events.
    pub async fn subscribe<F, Fut>(&self, handler: F)
    where
        F: Fn(DroneEvent) -> Fut + Send + Sync + 'static,
        Fut: std::future::Future<Output = ()> + Send + 'static,
    {
        self.handlers.write().await.push(Box::new(move |evt| Box::pin(handler(evt))));
    }

    /// Publish an event to all subscribers.
    pub async fn publish(&self, event: DroneEvent) {
        let handlers = self.handlers.read().await;
        for handler in handlers.iter() {
            handler(event.clone()).await;
        }
    }
}
